app/middleware/authentication.rb

```diff
@@ -11,25 +11,18 @@ def call(env)
       return unauthorized_response unless access_token
       decoded_access_token = JsonWebToken.new(access_token).verify!
       id_token = extract_id_token(env)
-
-      if id_token
-        id_token_json = JSON.parse(id_token)
-        email = id_token_json.dig('payload', 'email')
-      end
+      return unauthorized_response unless id_token
+      decoded_id_token = JsonWebToken.new(id_token).verify!
+      email = decoded_id_token.fetch('email')
 
       user = User.find_or_create_by(sub: decoded_access_token.fetch('sub')) do |u|
         u.email = email
         u.save!
       end
 
-      if user
-        if email
-          if user.email != email
-            user.email = email
-            user.save!
-          end
-        end
+      update_user_email_if_needed(user, email)
```

Comment:
We could move this down into the conditional below and not have to check the user twice, see my other comment :)